modules/effects/arrow-effects/src/main/kotlin/arrow/effects/typeclasses/Bracket.kt

```diff
@@ -0,0 +1,21 @@
+package arrow.effects.typeclasses
+
+import arrow.Kind
+import arrow.typeclasses.MonadError
+
+interface Bracket<F, E> : MonadError<F, E> {
+
+  fun <F, A, B> Kind<F, A>.bracketCase(release: (A, ExitCase<E>) -> Kind<F, Unit>, use: (A) -> Kind<F, B>): Kind<F, B>
+
+  fun <A, B> Kind<F, A>.bracket(release: (A) -> Kind<F, Unit>, use: (A) -> Kind<F, B>): Kind<F, B> =
+    bracketCase({ a, _ -> release(a) }, use)
+
+  fun <A> Kind<F, A>.uncancelable(): Kind<F, A> =
+    bracket({ just(Unit) }, { just(it) })
+
+  fun <A> Kind<F, A>.guarantee(finalizer: Kind<F, Unit>): Kind<F, A> =
```

Comment:
I was reading this code to learn more about typeclass implementation and I was left with a question. In [Cats docs](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Bracket.scala) `guarantee` has this signature and implementation:

`def guarantee[A](fa: F[A])(finalizer: F[Unit]): F[A] = bracket(unit)(_ => fa)(_ => finalizer)`

In this case, shouldn't signature use the receiver type `Kind<F, Unit>` instead of `Kind<F, A>`? Something like: 

`fun <A> Kind<F, Unit>.guarantee(finalizer: Kind<F, Unit>, fa: Kind<F, A>): Kind<F, A> = bracket({ _ -> finalizer }, { _ -> fa })`

Sorry about newbie level question and thanks in advance 😄

Reply:
Note that the Scala one gets an additional argument which is Fine<A>. That's what we move to be the receiver in Arrow.

Reply:
Yes, `guarantee` has `fa: F<A>`  as a parameter, but it call `bracket` passing `unit` as `acquire: F<A>` and `fa` as `use: A => F<B>`.

In Kotlin, aren't we using the receiver type as `acquire` param? If so, it looks like the receiver type `Kind<F, A>` are used in both `acquire` and `use` params of the `bracket` method.

Reply:
The method guarantee reads as follows: for the resource `A` provided, run `finalizer` when it completes. So your use would be as so:

```
val resource: Res = newRes()
just(resource).guarantee(invoke { resource.close() }) // Kind<F, A>
```

Which is a simpler version of:
```
val resource: Res = newRes()
bracket({ invoke { resource.close() } }) { just(it) } // Kind<F, A>
```

Reply:
Oh, now I got the idea. Reading that way makes totally sense the signature used!

Thank you both! :smile: